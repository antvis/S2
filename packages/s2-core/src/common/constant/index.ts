export * from './events';
export * from './basic';
export * from './classnames';
export * from './condition';
export * from './dataConfig';
export * from './frozen';
export * from './interaction';
export * from './options';
export * from './scroll';
export * from './theme';
export * from './tooltip';
export * from './resize';
