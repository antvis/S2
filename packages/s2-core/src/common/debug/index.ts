/* eslint-disable no-console */

export const DEBUG_TRANSFORM_DATA = 'Transform Data';
export const DEBUG_HEADER_LAYOUT = 'Header Layout';
export const DEBUG_VIEW_RENDER = 'Data Cell Render';

export class DebuggerUtil {
  private static instance: DebuggerUtil;

  private debug = false;

  public static getInstance() {
    if (!DebuggerUtil.instance) {
      DebuggerUtil.instance = new DebuggerUtil();
    }
    return DebuggerUtil.instance;
  }

  public setDebug(debug: boolean) {
    this.debug = debug;
  }

  public debugCallback = (info: string, callback: () => void) => {
    if (this.debug) {
      const start = performance.now();
      callback();
      const end = performance.now();
      console.log(info, `${end - start} ms`);
    } else {
      callback();
    }
  };

  public logger = (info: string, ...params: unknown[]) => {
    if (this.debug) {
      console.log(info, ...params);
    }
  };
}
