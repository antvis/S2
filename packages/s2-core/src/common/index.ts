export * from './constant';
export * from './debug';
export * from './i18n';
export * from './icons';
export * from './interface';
export * from './store';
