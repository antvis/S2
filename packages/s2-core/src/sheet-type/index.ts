import { PivotSheet } from './pivot-sheet';
import { TableSheet } from './table-sheet';
import { SpreadSheet } from './spread-sheet';

export { PivotSheet, TableSheet, SpreadSheet };
