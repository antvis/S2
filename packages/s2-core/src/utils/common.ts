export const safeJsonParse = <T = unknown>(val: string): T | null => {
  try {
    return JSON.parse(val);
  } catch (err) {
    return null;
  }
};
