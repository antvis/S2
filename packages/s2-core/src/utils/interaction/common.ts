import type { DisplayObject } from '@antv/g';
import type { CellAppendInfo } from '../../common';

export const getAppendInfo = <T extends Record<string, any> = CellAppendInfo>(
  eventTarget: DisplayObject,
) => {
  if (!eventTarget || !('appendInfo' in eventTarget)) {
    return {} as T;
  }

  return eventTarget.appendInfo as T;
};
