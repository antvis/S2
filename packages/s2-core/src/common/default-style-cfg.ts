import { TREE_ROW_DEFAULT_WIDTH } from './constant';
import { Style } from './interface';

/**
 * Create By Bruce Too
 * On 2020-11-27
 * 默认表单元格
 */
export const DefaultStyleCfg = () => {
  return {
    treeRowsWidth: TREE_ROW_DEFAULT_WIDTH,
    collapsedRows: {},
    collapsedCols: {},
    cellCfg: {
      width: 96,
      height: 30,
      padding: 0,
    },
    rowCfg: {
      width: 96,
      widthByField: {},
    },
    colCfg: {
      height: 40,
      widthByFieldValue: {},
      heightByField: {},
      colWidthType: 'adaptive',
      totalSample: 10,
      detailSample: 30,
      showDerivedIcon: true,
      maxSampleIndex: 1,
    },
    device: 'pc',
  } as Style;
};
