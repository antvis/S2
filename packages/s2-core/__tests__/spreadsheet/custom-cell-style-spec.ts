import {
  createPivotSheet,
  createTableSheet,
  getContainer,
} from 'tests/util/helpers';
import * as dataConfig from 'tests/data/mock-dataset.json';
import * as simpleDataConfig from 'tests/data/simple-data.json';
import { customColSimpleColumns } from '../data/custom-table-col-fields';
import { EXTRA_FIELD, type S2DataConfig } from '@/common';
import type { ViewMeta } from '@/common/interface/basic';
import type { Node } from '@/facet/layout/node';
import type { S2Options } from '@/common/interface';
import { TableSheet, type SpreadSheet, PivotSheet } from '@/sheet-type';

describe('SpreadSheet Custom Cell Style Tests', () => {
  let s2: SpreadSheet;

  const s2Options: S2Options = {
    width: 600,
    height: 400,
    hierarchyType: 'grid',
    transformCanvasConfig() {
      return {
        devicePixelRatio: 1,
      };
    },
  };

  const mapNodeSize = (nodes: Node[] | ViewMeta[]) =>
    nodes.map(({ id, width, height }) => {
      return { id, width, height };
    });

  describe('PivotSheet Custom Cell Style Tests', () => {
    beforeEach(async () => {
      s2 = createPivotSheet(s2Options);
      await s2.render();
    });

    afterEach(() => {
      s2.destroy();
    });

    test('should render default cell style', () => {
      expect(mapNodeSize(s2.facet.getRowNodes())).toMatchSnapshot();
      expect(mapNodeSize(s2.facet.getColNodes())).toMatchSnapshot();
      expect(
        mapNodeSize(s2.facet.getDataCells().map((cell) => cell.getMeta())),
      ).toMatchSnapshot();
    });

    describe('#DataCell', () => {
      test('should get custom data cell style', async () => {
        const sheet = createPivotSheet({
          ...s2Options,
          style: {
            dataCell: {
              width: 300,
              height: 90,
            },
          },
        });

        await sheet.render();

        expect(
          mapNodeSize(sheet.facet.getDataCells().map((cell) => cell.getMeta())),
        ).toMatchSnapshot();

        // 行头叶子节点高度和 dataCell 一致
        sheet.facet.getRowLeafNodes().forEach((node) => {
          expect(node.height).toStrictEqual(90);
        });

        // 列头叶子节点宽度和 dataCell 一致
        sheet.facet.getColLeafNodes().forEach((node) => {
          expect(node.width).toStrictEqual(300);
        });

        sheet.destroy();
      });
    });

    describe('#RowCell', () => {
      test('should get custom row cell style', async () => {
        s2.setOptions({
          style: {
            rowCell: {
              width: 50,
              height: 60,
            },
          },
        });
        await s2.render();

        expect(mapNodeSize(s2.facet.getRowNodes())).toMatchSnapshot();
      });

      test('should get custom row cell style by custom hook', async () => {
        s2.setOptions({
          style: {
            rowCell: {
              width: (node) => {
                expect(node?.id).toInclude('root[&]浙江');

                return 100;
              },
              height: (node) => {
                expect(node?.id).toInclude('root[&]浙江');

                return 40;
              },
            },
          },
        });
        await s2.render();

        expect(mapNodeSize(s2.facet.getRowNodes())).toMatchSnapshot();
      });

      test('should get custom row cell style by field', async () => {
        s2.setOptions({
          style: {
            rowCell: {
              widthByField: {
                city: 66,
              },
              heightByField: {
                city: 33,
              },
            },
          },
        });
        await s2.render();

        expect(mapNodeSize(s2.facet.getRowNodes())).toMatchSnapshot();
      });

      test('should get custom row cell style by field id', async () => {
        const cityId = `root[&]浙江[&]杭州`;

        s2.setOptions({
          style: {
            rowCell: {
              widthByField: {
                [cityId]: 50,
              },
              heightByField: {
                [cityId]: 300,
              },
            },
          },
        });
        await s2.render();

        expect(mapNodeSize(s2.facet.getRowNodes())).toMatchSnapshot();
      });

      test('should not get custom row cell style if not leaf node by field', async () => {
        const width = 100;
        const height = 200;

        const provinceId = `root[&]浙江`;

        s2.setOptions({
          style: {
            rowCell: {
              widthByField: {
                province: width,
                [provinceId]: width,
              },
              heightByField: {
                province: height,
                [provinceId]: height,
              },
            },
          },
        });
        await s2.render();

        const rootRowNodes = s2.facet
          .getRowNodes()
          .filter((node) => node.level === 0);

        expect(rootRowNodes).toHaveLength(1);
        rootRowNodes.forEach((node) => {
          expect(node.width).toStrictEqual(width);
          expect(node.height).not.toStrictEqual(height);
        });
      });

      test('should get custom row cell style priority by rowCell.heightByField > rowCell.height > dataCell.height', async () => {
        const sheet = createPivotSheet({
          ...s2Options,
          style: {
            rowCell: {
              heightByField: {
                city: 40,
              },
              height: 20,
            },
            dataCell: {
              height: 50,
            },
          },
        });

        await sheet.render();

        // 1. rowCell.heightByField > rowCell.height > dataCell.height
        sheet.facet.getRowLeafNodes().forEach((node) => {
          expect(node.height).toEqual(40);
        });

        // 2. rowCell.height > dataCell.height
        sheet.setOptions(
          {
            style: {
              rowCell: {
                height: 20,
              },
              dataCell: {
                height: 50,
              },
            },
          },
          true,
        );
        await sheet.render(false);

        sheet.facet.getRowLeafNodes().forEach((node) => {
          expect(node.height).toEqual(20);
        });
      });
    });

    describe('#ColCell', () => {
      test('should get custom col cell style', async () => {
        s2.setOptions({
          style: {
            colCell: {
              width: 50,
              height: 60,
            },
          },
        });
        await s2.render();

        expect(mapNodeSize(s2.facet.getColNodes())).toMatchSnapshot();
      });

      test('should get custom col cell style by custom hook', async () => {
        s2.setOptions({
          style: {
            colCell: {
              width: (node) => {
                expect(node?.id).toInclude('root[&]笔');

                return 100;
              },
              height: (node) => {
                expect(node?.id).toInclude('root[&]笔');

                return 40;
              },
            },
          },
        });
        await s2.render();

        expect(mapNodeSize(s2.facet.getColNodes())).toMatchSnapshot();
      });

      test('should get custom col cell style by field', async () => {
        s2.setOptions({
          style: {
            colCell: {
              widthByField: {
                sub_type: 66,
              },
              heightByField: {
                sub_type: 33,
              },
            },
          },
        });
        await s2.render();

        expect(mapNodeSize(s2.facet.getColNodes())).toMatchSnapshot();
      });

      test('should get custom col cell style by field id', async () => {
        const priceId = `root[&]笔[&]price`;

        s2.setOptions({
          style: {
            colCell: {
              widthByField: {
                [priceId]: 50,
              },
              heightByField: {
                [priceId]: 300,
              },
            },
          },
        });
        await s2.render();

        expect(mapNodeSize(s2.facet.getRowNodes())).toMatchSnapshot();
      });

      test('should not get custom col cell style if not leaf node by field', async () => {
        const width = 100;
        const height = 200;

        const subTypeId = `root[&]笔[&]sub_type`;

        s2.setOptions({
          style: {
            colCell: {
              widthByField: {
                sub_type: width,
                [subTypeId]: width,
              },
              heightByField: {
                sub_type: height,
                [subTypeId]: height,
              },
            },
          },
        });
        await s2.render();

        const rootColNodes = s2.facet
          .getColNodes()
          .filter((node) => node.level === 0);

        expect(rootColNodes).toHaveLength(1);
        rootColNodes.forEach((node) => {
          expect(node.width).not.toStrictEqual(width);
          expect(node.height).not.toStrictEqual(height);
        });
      });

      test('should get custom col cell style priority by colCell.widthByField > colCell.width > dataCell.width', async () => {
        const sheet = createPivotSheet({
          ...s2Options,
          style: {
            colCell: {
              width: 200,
              widthByField: {
                [EXTRA_FIELD]: 300,
              },
            },
            dataCell: {
              width: 100,
            },
          },
        });

        await sheet.render();

        // 1. colCell.widthByField > colCell.width > dataCell.width
        sheet.facet.getColLeafNodes().forEach((node) => {
          expect(node.width).toEqual(300);
        });

        // 2. colCell.width > dataCell.width
        sheet.setOptions(
          {
            style: {
              colCell: {
                width: 200,
              },
              dataCell: {
                width: 100,
              },
            },
          },
          true,
        );
        await sheet.render(false);

        sheet.facet.getColLeafNodes().forEach((node) => {
          expect(node.width).toEqual(200);
        });
      });

      test('should get custom col cell style if measure column hidden', async () => {
        const sheet = new PivotSheet(
          getContainer(),
          {
            ...simpleDataConfig,
            fields: {
              ...simpleDataConfig.fields,
              values: ['price'],
            },
          },
          {
            ...s2Options,
            style: {
              colCell: {
                hideValue: true,
                widthByField: {
                  'root[&]笔': 100,
                },
              },
            },
          },
        );

        await sheet.render();

        sheet.facet.getColLeafNodes().forEach((node) => {
          expect(node.width).toEqual(100);
        });
      });
    });
  });

  describe('TableSheet Custom Cell Style Tests', () => {
    beforeEach(async () => {
      s2 = createTableSheet(s2Options, {
        useSimpleData: false,
      });
      await s2.render();
    });

    afterEach(() => {
      s2.destroy();
    });

    test('should render default cell style', () => {
      expect(mapNodeSize(s2.facet.getRowNodes())).toMatchSnapshot();
      expect(mapNodeSize(s2.facet.getColNodes())).toMatchSnapshot();
      expect(
        mapNodeSize(s2.facet.getDataCells().map((cell) => cell.getMeta())),
      ).toMatchSnapshot();
    });

    test('should set row cell style by height', async () => {
      s2.setOptions({
        style: {
          rowCell: {
            height: 60,
          },
        },
      });
      await s2.render();

      expect(
        mapNodeSize(s2.facet.getDataCells().map((cell) => cell.getMeta())),
      ).toMatchSnapshot();
    });

    test('should set row cell style by row index', async () => {
      s2.setOptions({
        style: {
          rowCell: {
            heightByField: {
              0: 40,
              '2': 100,
            },
          },
        },
      });
      await s2.render();

      expect(
        mapNodeSize(s2.facet.getDataCells().map((cell) => cell.getMeta())),
      ).toMatchSnapshot();
    });

    test('should set col cell style by width', async () => {
      s2.setOptions({
        style: {
          colCell: {
            width: 100,
          },
        },
      });
      await s2.render();

      expect(mapNodeSize(s2.facet.getColNodes())).toMatchSnapshot();
    });

    test('should set col cell style by field', async () => {
      const sheet = createTableSheet(s2Options, {
        useSimpleData: false,
      });

      sheet.setOptions({
        style: {
          colCell: {
            widthByField: {
              type: 50,
              sub_type: 120,
            },
          },
        },
      });
      await sheet.render();

      expect(mapNodeSize(sheet.facet.getColNodes())).toMatchSnapshot();
    });

    test('should set col cell style by field id', async () => {
      const sheet = createTableSheet(s2Options, {
        useSimpleData: false,
      });

      sheet.setOptions({
        style: {
          colCell: {
            widthByField: {
              'root[&]类别': 100,
              'root[&]子类别': 200,
            },
          },
        },
      });
      await sheet.render();

      expect(mapNodeSize(sheet.facet.getColNodes())).toMatchSnapshot();
    });

    test('should not set top level col cell style by field', async () => {
      const customColDataCfg: S2DataConfig = {
        ...dataConfig,
        fields: {
          columns: customColSimpleColumns,
        },
      };
      const sheet = new TableSheet(getContainer(), customColDataCfg, s2Options);

      sheet.setOptions({
        style: {
          colCell: {
            widthByField: {
              area: 100,
              money: 100,
            },
          },
        },
      });
      await sheet.render();

      const nodes = sheet.facet.getColNodes();

      const areaNode = nodes.find((node) => node.field === 'area');
      const moneyNode = nodes.find((node) => node.field === 'money');

      expect(areaNode?.width).not.toEqual(100);
      expect(moneyNode?.width).not.toEqual(100);
    });

    test('should set col cell style by field for multiple columns', async () => {
      const customColDataCfg: S2DataConfig = {
        ...dataConfig,
        fields: {
          columns: customColSimpleColumns,
        },
      };
      const sheet = new TableSheet(getContainer(), customColDataCfg, s2Options);

      sheet.setOptions({
        style: {
          colCell: {
            widthByField: {
              province: 100,
              city: 80,
              price: 90,
              number: 120,
            },
          },
        },
      });
      await sheet.render();

      expect(mapNodeSize(sheet.facet.getColNodes())).toMatchSnapshot();
    });
  });
});
