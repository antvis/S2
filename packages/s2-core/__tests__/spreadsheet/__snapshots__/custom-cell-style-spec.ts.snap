// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpreadSheet Custom Cell Style Tests PivotSheet Custom Cell Style Tests #ColCell should get custom col cell style 1`] = `
Array [
  Object {
    "height": 60,
    "id": "root[&]笔",
    "width": 100,
  },
  Object {
    "height": 60,
    "id": "root[&]笔[&]price",
    "width": 50,
  },
  Object {
    "height": 60,
    "id": "root[&]笔[&]cost",
    "width": 50,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests PivotSheet Custom Cell Style Tests #ColCell should get custom col cell style by custom hook 1`] = `
Array [
  Object {
    "height": 40,
    "id": "root[&]笔",
    "width": 200,
  },
  Object {
    "height": 40,
    "id": "root[&]笔[&]price",
    "width": 100,
  },
  Object {
    "height": 40,
    "id": "root[&]笔[&]cost",
    "width": 100,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests PivotSheet Custom Cell Style Tests #ColCell should get custom col cell style by field 1`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]笔",
    "width": 299,
  },
  Object {
    "height": 30,
    "id": "root[&]笔[&]price",
    "width": 149.5,
  },
  Object {
    "height": 30,
    "id": "root[&]笔[&]cost",
    "width": 149.5,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests PivotSheet Custom Cell Style Tests #ColCell should get custom col cell style by field id 1`] = `
Array [
  Object {
    "height": 60,
    "id": "root[&]浙江",
    "width": 149.5,
  },
  Object {
    "height": 30,
    "id": "root[&]浙江[&]义乌",
    "width": 149.5,
  },
  Object {
    "height": 30,
    "id": "root[&]浙江[&]杭州",
    "width": 149.5,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests PivotSheet Custom Cell Style Tests #DataCell should get custom data cell style 1`] = `
Array [
  Object {
    "height": 90,
    "id": "root[&]浙江[&]义乌-root[&]笔[&]price",
    "width": 300,
  },
  Object {
    "height": 90,
    "id": "root[&]浙江[&]杭州-root[&]笔[&]price",
    "width": 300,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests PivotSheet Custom Cell Style Tests #RowCell should get custom row cell style 1`] = `
Array [
  Object {
    "height": 120,
    "id": "root[&]浙江",
    "width": 50,
  },
  Object {
    "height": 60,
    "id": "root[&]浙江[&]义乌",
    "width": 50,
  },
  Object {
    "height": 60,
    "id": "root[&]浙江[&]杭州",
    "width": 50,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests PivotSheet Custom Cell Style Tests #RowCell should get custom row cell style by custom hook 1`] = `
Array [
  Object {
    "height": 80,
    "id": "root[&]浙江",
    "width": 100,
  },
  Object {
    "height": 40,
    "id": "root[&]浙江[&]义乌",
    "width": 100,
  },
  Object {
    "height": 40,
    "id": "root[&]浙江[&]杭州",
    "width": 100,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests PivotSheet Custom Cell Style Tests #RowCell should get custom row cell style by field 1`] = `
Array [
  Object {
    "height": 66,
    "id": "root[&]浙江",
    "width": 149.5,
  },
  Object {
    "height": 33,
    "id": "root[&]浙江[&]义乌",
    "width": 66,
  },
  Object {
    "height": 33,
    "id": "root[&]浙江[&]杭州",
    "width": 66,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests PivotSheet Custom Cell Style Tests #RowCell should get custom row cell style by field id 1`] = `
Array [
  Object {
    "height": 330,
    "id": "root[&]浙江",
    "width": 149.5,
  },
  Object {
    "height": 30,
    "id": "root[&]浙江[&]义乌",
    "width": 149.5,
  },
  Object {
    "height": 300,
    "id": "root[&]浙江[&]杭州",
    "width": 149.5,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests PivotSheet Custom Cell Style Tests should render default cell style 1`] = `
Array [
  Object {
    "height": 60,
    "id": "root[&]浙江",
    "width": 149.5,
  },
  Object {
    "height": 30,
    "id": "root[&]浙江[&]义乌",
    "width": 149.5,
  },
  Object {
    "height": 30,
    "id": "root[&]浙江[&]杭州",
    "width": 149.5,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests PivotSheet Custom Cell Style Tests should render default cell style 2`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]笔",
    "width": 299,
  },
  Object {
    "height": 30,
    "id": "root[&]笔[&]price",
    "width": 149.5,
  },
  Object {
    "height": 30,
    "id": "root[&]笔[&]cost",
    "width": 149.5,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests PivotSheet Custom Cell Style Tests should render default cell style 3`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]浙江[&]义乌-root[&]笔[&]price",
    "width": 149.5,
  },
  Object {
    "height": 30,
    "id": "root[&]浙江[&]杭州-root[&]笔[&]price",
    "width": 149.5,
  },
  Object {
    "height": 30,
    "id": "root[&]浙江[&]义乌-root[&]笔[&]cost",
    "width": 149.5,
  },
  Object {
    "height": 30,
    "id": "root[&]浙江[&]杭州-root[&]笔[&]cost",
    "width": 149.5,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests TableSheet Custom Cell Style Tests should render default cell style 1`] = `Array []`;

exports[`SpreadSheet Custom Cell Style Tests TableSheet Custom Cell Style Tests should render default cell style 2`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "root[&]sub_type",
    "width": 299.5,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests TableSheet Custom Cell Style Tests should render default cell style 3`] = `
Array [
  Object {
    "height": 30,
    "id": "0-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "1-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "2-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "3-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "4-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "5-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "6-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "7-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "8-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "9-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "10-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "11-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "12-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "0-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "1-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "2-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "3-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "4-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "5-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "6-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "7-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "8-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "9-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "10-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "11-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "12-root[&]sub_type",
    "width": 299.5,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests TableSheet Custom Cell Style Tests should set col cell style by field 1`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]type",
    "width": 50,
  },
  Object {
    "height": 30,
    "id": "root[&]sub_type",
    "width": 120,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests TableSheet Custom Cell Style Tests should set col cell style by field for multiple columns 1`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]area",
    "width": 180,
  },
  Object {
    "height": 30,
    "id": "root[&]area[&]province",
    "width": 100,
  },
  Object {
    "height": 30,
    "id": "root[&]area[&]city",
    "width": 80,
  },
  Object {
    "height": 60,
    "id": "root[&]type",
    "width": 119.8,
  },
  Object {
    "height": 30,
    "id": "root[&]money",
    "width": 210,
  },
  Object {
    "height": 30,
    "id": "root[&]money[&]price",
    "width": 90,
  },
  Object {
    "height": 30,
    "id": "root[&]money[&]number",
    "width": 120,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests TableSheet Custom Cell Style Tests should set col cell style by field id 1`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "root[&]sub_type",
    "width": 299.5,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests TableSheet Custom Cell Style Tests should set col cell style by width 1`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]type",
    "width": 100,
  },
  Object {
    "height": 30,
    "id": "root[&]sub_type",
    "width": 100,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests TableSheet Custom Cell Style Tests should set row cell style by height 1`] = `
Array [
  Object {
    "height": 60,
    "id": "0-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 60,
    "id": "1-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 60,
    "id": "2-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 60,
    "id": "3-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 60,
    "id": "4-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 60,
    "id": "5-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 60,
    "id": "6-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 60,
    "id": "0-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 60,
    "id": "1-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 60,
    "id": "2-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 60,
    "id": "3-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 60,
    "id": "4-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 60,
    "id": "5-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 60,
    "id": "6-root[&]sub_type",
    "width": 299.5,
  },
]
`;

exports[`SpreadSheet Custom Cell Style Tests TableSheet Custom Cell Style Tests should set row cell style by row index 1`] = `
Array [
  Object {
    "height": 40,
    "id": "0-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "1-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 100,
    "id": "2-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "3-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "4-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "5-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "6-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "7-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "8-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "9-root[&]type",
    "width": 299.5,
  },
  Object {
    "height": 40,
    "id": "0-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "1-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 100,
    "id": "2-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "3-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "4-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "5-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "6-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "7-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "8-root[&]sub_type",
    "width": 299.5,
  },
  Object {
    "height": 30,
    "id": "9-root[&]sub_type",
    "width": 299.5,
  },
]
`;
